### PR TITLE
INTG-174 csv export

### PR DESCRIPTION
**Pull request by tonyOreglia**

### CODE REVIEW

File: tools/exporter/exporter.py [hunk 1]
```diff
@@ -470,6 +469,12 @@ def sync_exports(logger, sc_client, settings):
                     export_doc = sc_client.get_export(audit_id, timezone, export_profile_id, export_format)
                 elif export_format == 'json':
                     export_doc = json.dumps(audit_json, indent=4)
+                elif export_format == 'csv':
+                    csv_exporter = csv.CsvExporter(audit_json)
+                    export_filename = audit_json['template_id']
+                    csv_exporter.write(export_path, export_filename, 'ab')
+                    # skip call to save_exported_document, csv_exporter class handles it with call to write()
```
Comment (dimitrikoubaroulis) [inline tools/exporter/exporter.py]: Let's make this consistent and have a function that returns the export_doc here.

File: tools/exporter/tests/test_csv_exporter.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import sys
+import unittest
+import json
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'exporter'))
+import csvExporter as csv
+
+class ExporterTestCase(unittest.TestCase):
+
+    def test_unit_one(self):
```
Comment (dimitrikoubaroulis) [inline tools/exporter/tests/test_csv_exporter.py]: def test_convert_single_question_with_standard_yes_no_na_answered_yes(...):
and remove the comment
also rename the expected output file, hard to understand what test_1 means.

File: tools/exporter/tests/test_csv_exporter.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        # single question field with standard yes/no/na answer set, answered 'yes'
+        csv_exporter = csv.CsvExporter(json.load(open('csv_test_files/unit_test_1_audit.json', 'r')))
+        self.assertEqual(csv.add_header(csv_exporter.process_items()), open('csv_test_files/unit_test_1_expected_output.csv', 'r').read())
+
+
+if __name__ == '__main__':
+        unittest.main()
```
Comment (dimitrikoubaroulis) [inline tools/exporter/tests/test_csv_exporter.py]: It's good practice to always have a newline at the end of every source file

File: tools/exporter/csvExporter.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+
+class CsvExporter:
+    """
+    provides tools to convert single json audit to CSV
+
+    Attributes:
+        audit_json(json):       audit to be converted to csv
+        audit_id (str):         id of current audit
+        auditdata (json):       general audit information
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        items (json):           list of audit's data input fields
+        auditdata_array(list):  list of general audit information
+        data(list):             2 diminsional list, each list is a single item, which corresponds to a single row
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        """
+        initialize class
+
+        :param audit_json:      audit in json format to be converted to csv
+        """
+        self.audit_json = audit_json
+        self.audit_id = audit_json['audit_id']
+        audit_data_and_items = self.get_items_and_auditdata()
+        self.auditdata = audit_data_and_items[0]
+        self.items = audit_data_and_items[1]
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+        self.auditdata_array = self.retrieve_auditdata()
+        self.data = self.process_items()
+
+    def process_items(self):
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        """
+        convert json of audit items to 2 dimensional list
+
+        :return:    2 diminsional list, each list is a single item, which corresponds to a single row
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: doh! ...replaced

File: tools/exporter/csvExporter.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+        """
+        self.data = []
+        for item in self.items:
+            item_array = self.generate_csv_row_item_data(item)
+            row_array = item_array + self.auditdata_array
+            self.data.append(row_array)
+        return self.data
+
+    def write(self, path, filename, write_or_append='wb'):
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+        """
+        write audit data to csv, this can be run immediately after instance of class is initialized
+
+        :param path:            path to desired output location
+        :param filename:        desired name of file
+        :param write_or_append:
+        :return:
+        """
+        if filename[-4:] != '.csv':
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]:
```
if not filename.lower().endswith('.csv'):
    filename = filename + '.csv'
```

File: tools/exporter/csvExporter.py [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+        if not os.path.isfile(file_path) or write_or_append == 'wb':
+            self.data.insert(0, CSV_HEADER_ROW)
+        try:
+            csv_file = open(file_path, write_or_append)
+            wr = csv.writer(csv_file, quoting=csv.QUOTE_ALL)
+            for row in self.data:
+                wr.writerow(row)
+            csv_file.close()
+        except Exception as ex:
+            print str(ex) + ': Exception while writing' + filename + ' to file'
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+
+    def path(self, obj, *args):
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+        """
+        Check for the existence of path for a given JSON object
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]:
```
Returns json value, if it exists. If it does not exist, this fxn returnsan empty string
```

File: tools/exporter/csvExporter.py [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+
+    # noinspection PyDictCreation
+    def generate_csv_row_item_data(self, item):
+        """
+        Process item JSON data into array that CSV writer can handle
+        :param item:    single item in JSON format
+        :return:        array of item data, in format that CSV writer can handle
+        """
+
+        fields = {
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated variable name to item_values. Is a plural variable name appropriate when it carries multiple item values?

File: tools/exporter/csvExporter.py [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+            FAILED: BLANK,
+            SCORE: BLANK,
+            MAX_SCORE: BLANK,
+            SCORE_PERCENTAGE: BLANK,
+            PARENT_ID: BLANK,
+            RESPONSE: BLANK
+        }
+
+        # all fields have a parent id, grab this and insert into fields dictionary
+        fields[PARENT_ID] = self.path(item, 'parent_id')
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: updated

File: tools/exporter/csvExporter.py [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+        # if failed key exists, insert value into dictionary
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: path (now called get_json_property) returns an empty string if value does not exist. So essentially leave the value empty. Comment updated to 
```
# insert value into dictionary, get_json_property returns empty string if no value exists
```

File: tools/exporter/csvExporter.py [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+        fields[FAILED] = self.path(item, 'responses', 'failed')
+
+        # check if this item has score information. insert data into dictionary
```
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: OK, makes sense. updated

File: tools/exporter/csvExporter.py (hunk 16 above)
Comment (tonyOreglia) [inline tools/exporter/csvExporter.py]: removed comment as there is not an unclear 'why?' question regarding this line of code

**Comment by tonyOreglia** [comment]
updated per Dimitri's Code Review, and Pair Programming feedback.

File: tools/exporter/csvExporter.py [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+                response = response[1:]
+        elif item.get('type') == 'signature':
+            response = self.get_json_property(item, 'responses', 'image', 'media_id')
+        elif item.get('type') == 'smartfield':
+            response = self.get_json_property(item, 'evaluation')
+        elif item.get('type') == 'datetime':
+            response = self.get_json_property(item, 'responses', 'datetime')
+        elif item.get('type') in ['dynamicfield', 'element', 'primeelement', 'asset', 'scanner', 'category', 'text',
+                                  'textsingle', 'section', 'information']:
+            pass
```
Comment (dimitrikoubaroulis) [inline tools/exporter/csvExporter.py]: Add a comment explaining why these are 'pass'-ed.

File: tools/exporter/csvExporter.py [hunk 19]
```diff
@@ -114,9 +205,12 @@ def save_converted_audit_to_file(self, output_csv_path, allow_overwrite):
...
         :param output_csv_path:  The full path to the file to save
         """
-        if os.path.isfile(output_csv_path) and not allow_overwrite:
-            sys.exit('File already exists at ' + output_csv_path + '\nPlease set allow_overwrite to True in confil.yaml file. See ReadMe.md for further instruction')
-        elif os.path.isfile(output_csv_path) and allow_overwrite:
+        file_exists = os.path.isfile(output_csv_path)
+        if file_exists and not allow_overwrite:
+            sys.exit(
+                'File already exists at ' + output_csv_path +
+                '\nPlease set allow_overwrite to True in confil.yaml file. See ReadMe.md for further instruction')
```
Comment (dimitrikoubaroulis) [inline tools/exporter/csvExporter.py]: confil => config. Remove the part about ReadMe as it is implied.

File: tools/exporter/csvExporter.py [hunk 20]
```diff
@@ -114,9 +205,12 @@ def save_converted_audit_to_file(self, output_csv_path, allow_overwrite):
 
         :param output_csv_path:  The full path to the file to save
         """
-        if os.path.isfile(output_csv_path) and not allow_overwrite:
-            sys.exit('File already exists at ' + output_csv_path + '\nPlease set allow_overwrite to True in confil.yaml file. See ReadMe.md for further instruction')
-        elif os.path.isfile(output_csv_path) and allow_overwrite:
+        file_exists = os.path.isfile(output_csv_path)
+        if file_exists and not allow_overwrite:
+            sys.exit(
```
Comment (dimitrikoubaroulis) [inline tools/exporter/csvExporter.py]: sys.exit() is a bit harsh here. I suggest simply skip saving the file if it exists.

File: tools/exporter/csvExporter.py [hunk 21]
```diff
@@ -68,23 +135,47 @@ def audit_items(self):
...
+        audit_data_as_list.append(self.reformat_date(audit_data_property['date_completed']))
         audit_data_as_list.append(self.audit_id())
+        audit_data_as_list.append(self.audit_json['template_id'])
+        audit_data_as_list.append(template_data_property['metadata']['name'])
+        audit_data_as_list.append(template_data_property['authorship']['author'])
         return audit_data_as_list
 
+    def reformat_date(self, date):
+        """
+        :param date:    date in the format: 2017-03-03T03:45:58.090Z
```
Comment (dimitrikoubaroulis) [inline tools/exporter/csvExporter.py]: A comment about 'why' we reformat would help here. Or name the function convert_to_user_friendly_date(self, iso_date).

File: tools/exporter/csvExporter.py [hunk 22]
```diff
@@ -2,40 +2,107 @@
...
+    '6ff300f0-e501-11e1-aff1-0800200c9a66': 'text is',
+    '6ff300f1-e501-11e1-aff1-0800200c9a66': 'text is not',
+    # conditional statements for signature item
+    '831f8ff0-e500-11e1-aff1-0800200c9a66': 'if signature exists',
+    '831f8ff1-e500-11e1-aff1-0800200c9a66': 'if the signature does not exist',
+    '831f8ff2-e500-11e1-aff1-0800200c9a66': 'if the signature name is',
+    '831f8ff3-e500-11e1-aff1-0800200c9a66': 'if the signature name is not'
+}
+
+standard_response_id_map = {
```
Comment (dimitrikoubaroulis) [inline tools/exporter/csvExporter.py]: standard => default (to be consistent with how we refer to these elsewhere)

File: tools/exporter/ReadMe.md [hunk 23]
```diff
@@ -45,11 +45,13 @@ Note:
...
 ### CSV Export
-* Audits that are build off a common Template will appended one after the other into the same CSV file. 
+#### Bulk Export Mode
+* Audits that are built off a common Template will be appended one after the other into the same CSV file. 
 * For an example of the CSV format, open the file: 
 ```
 safetyculture-sdk-python/tools/exporter/tests/csv_test_files/test_convert_single_question_with_standard_yes_no_na_answered_yes_expected_output.csv
 ```
+#### Single Audit Export Mode
 * To export a single Audit, first export the Audit in JSON format, then call csvExporter.py with the Audit JSON sent as an argument:
```
Comment (dimitrikoubaroulis) [inline tools/exporter/ReadMe.md]: first save the audit JSON to a file, then pass this file as an argument to csvExporter.py, like this:

File: tools/exporter/ReadMe.md [hunk 24]
```diff
@@ -45,11 +45,13 @@ Note:
 * Up to 1000 audits will be exported each time the software checks for new audits. If more than 1000 audits exist on the SafetyCulture platform, they will be retrieved automatically in subsequent sync cycles.
 
 ### CSV Export
-* Audits that are build off a common Template will appended one after the other into the same CSV file. 
+#### Bulk Export Mode
+* Audits that are built off a common Template will be appended one after the other into the same CSV file.
```
Comment (dimitrikoubaroulis) [inline tools/exporter/ReadMe.md]: Audits conducted against the same template will be appended to a single CSV file named after the template ID (or name, whatever the case is).

**Pull request merged**